src/main/scala/edu/ie3/util/osm/model/OsmEntity.scala

```diff
@@ -0,0 +1,162 @@
+  final def hasKeysValuesPairOr(
+      keyTagValues: Map[String, Set[String]]
+  ): Boolean = {
+    keyTagValues.headOption match {
+      case Some((key, tagValues)) =>
+        // if we have at least one matching key/tagValues pair
+        // we can stop the process and return true
+        if (hasKeyValuesPairOr(key, tagValues)) {
+          true
+        } else {
+          hasKeysValuesPairOr(keyTagValues - key)
+        }
+      case None =>
+        false
+    }
+  }
+
+  def hasKey(key: CommonOsmKey): Boolean =
+    hasKeyValuesPairOr(key, Set.empty)
+
+  def hasKey(key: String): Boolean =
+    hasKeyValuesPairOr(key, Set.empty)
+
+}
+
+object OsmEntity {
+
+  enum OsmEntityType:
+    case Node, Way, Relation
+
+  final case class MetaInformation(
+      version: Option[Int] = None,
+      timestamp: Option[Instant] = None,
+      changeSet: Option[Long] = None,
+      userId: Option[Int] = None,
+      userName: Option[String] = None,
+      visible: Option[Boolean] = None
+  )
+
+  final case class Node(
+      id: Long,
+      latitude: Double,
+      longitude: Double,
+      tags: Map[String, String],
+      metaInformation: Option[MetaInformation] = None
+  ) extends OsmEntity {
+    override val osmModel: OsmEntityType = OsmEntityType.Node
+    val coord: Point = GeoUtils.xyToPoint(longitude, latitude)
+  }
+
+  sealed trait Way extends OsmEntity {
+    val id: Long
+    val nodes: Seq[Long]
+    val tags: Map[String, String]
+    val metaInformation: Option[MetaInformation]
+
+    override val osmModel: OsmEntityType = OsmEntityType.Way
+  }
+
+  object Way {
+
+    final case class OpenWay private[model] (
+        id: Long,
+        nodes: Seq[Long],
+        tags: Map[String, String],
+        metaInformation: Option[MetaInformation]
+    ) extends Way
+
+    final case class ClosedWay private[model] (
+        id: Long,
+        nodes: Seq[Long],
+        tags: Map[String, String],
+        metaInformation: Option[MetaInformation]
+    ) extends Way
+
+    def apply(
+        id: Long,
+        nodes: Seq[Long],
+        tags: Map[String, String],
+        metaInformation: Option[MetaInformation]
+    ): Way =
+      if (closedWay(nodes)) {
+        ClosedWay(id, nodes, tags, metaInformation)
+      } else {
+        OpenWay(id, nodes, tags, metaInformation)
+      }
+
+    def closedWay(nodes: Seq[Long]): Boolean =
+      nodes.headOption.zip(nodes.lastOption).exists { case (head, last) =>
+        head == last
+      }
+
+  }
+
+  final case class Relation(
+      id: Long,
+      relations: Seq[RelationMember],
+      tags: Map[String, String],
+      metaInformation: Option[MetaInformation] = None
+  ) extends OsmEntity {
+    override val osmModel: OsmEntityType = OsmEntityType.Relation
+  }
+
+  object Relation {
+
+    enum RelationMemberType:
+      case Node, Way, Relation, Unrecognized
+
+    final case class RelationMember(
+        id: Long,
+        relationTypes: RelationMemberType,
+        role: String
```

Comment:
I'd find a short ScalaDoc helpful (maybe with a reference to OpenStreetMap wiki) to understand the role of `role`.